### PR TITLE
[Performance] Optimize AdamW GPU kernel and Change the learning rate type to float64

### PR DESCRIPTION
**Pull request by zhengshengning**
### PR Category                                                                                                                                                                                                                                  
  Performance Optimization                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                   
  ### PR Types                                                                                                                                                                                                                                     
  Improvements                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                   
  ### Description                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                   
  #### 背景                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                 
  Paddle AdamW / fused_adam 优化器长期以 **float32** 精度存储学习率（LR），导致：                                                                                                                                                                  
                                                                                                                                                                                                                                                 
  1. **精度损失**：float32 仅有约 7 位有效数字，当 LR 很小（如 `1e-6`）时，float32 截断会引入不可忽视的误差。                                                                                                                                                                                                                    
  2. **Python 侧精度补偿 hack**：为弥补 float32 精度损失，`adamw.py` 通过调整 `lr_ratio = lr_ratio * (double_lr / float32_lr)` 绕开问题，增加了维护负担。                                                                                          
                                                                                                                                                                                                                                                   
  #### 核心改动                                                                                                                                                                                                                                    
                                                                                                                                                                                                                      
  - `Optimizer.get_lr_dtype()` 重构为独立方法，`AdamW` 覆盖后固定返回 `paddle.float64`。
  - GPU / CPU kernel 的 LR 输入注册 `SetDataType(FLOAT64)`；CUDA kernel 内部 `lr_` 指针从 `const MT*` 改为 `const double*`。                                                                                                                       
  - XPU kernel 在读取 LR 前自动 cast `float64 → float32`（XPU XDNN 仅支持 float32）。                                       
  - 删除 `adamw.py` 中 float32 精度补偿逻辑。                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                                                               
                                                                                                                                    
  **保留 PyTorch 兼容精度模式**                                                                                                                                                                                                                 
                                           
  通过 `FLAGS_use_accuracy_compatible_kernel=true` 可开启兼容模式，通过 `log(beta_pow)/log(beta)` 反推 step_count 的计算路径，以实现与 PyTorch bit-for-bit 数值对齐。                                                                      
                                                                                                                                                                             
  #### 涉及文件                                                                                                                                                                                                                                    
                                           
  | 文件 | 变更说明 |                                                                                                                                                                                                                              
  |---|---|                                
  | `paddle/phi/kernels/gpu/adamw_kernel.cu` | 核心重构：float64 LR、LrAccessor、BiasCorrAccessor、消除 GPU sync |                                                                                                                                 
  | `paddle/phi/kernels/gpu/fused_adam_kernel.cu` | LR 改为 `double*` |                                           
  | `paddle/phi/kernels/gpu/adam_kernel.cu` | `MPTypeTrait` 命名空间清理 |                                                                                                                                                                         
  | `paddle/phi/kernels/cpu/adamw_kernel.cc` | float64 LR 支持 |                                                                                                                                                                                   
  | `paddle/phi/kernels/cpu/fused_adam_kernel.cc` | float64 LR + cast |                                                                                                                                                                            
  | `paddle/phi/kernels/xpu/adamw_kernel.cc` | float64 LR → float32 cast for XPU |                                                                                                                                                                 
  | `python/paddle/optimizer/optimizer.py` | 将 lr dtype 逻辑重构为 `get_lr_dtype()` 方法 |                                                                                                                                                        
  | `python/paddle/optimizer/adamw.py` | 覆盖 `get_lr_dtype()` 返回 float64，删除 lr_ratio hack |                                                                                                                                                  
  | `test/legacy_test/test_adamw_op.py` | LR 测试数据改为 float64 |                                                                                                                                                                                
  | `test/legacy_test/test_fused_adam_op.py` | 同上 |                                                                                                                                                                                              
  | `test/collective/fleet/hybrid_parallel_sharding_muon_model.py` | Muon sharding 测试更新 |                                                                                                                                                      
                                                                                                                                                                                                                                                   
  ### 是否引起精度变化                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                   
  是。学习率由 float32 升级为 float64，精度从约 7 位有效数字提升至约 15 位，对原来的精度进行了提升。当FLAGS_use_accuracy_compatible_kernel=false时，走的是paddle原来的实现版本（只有lr的类型变成了double）。FLAGS_use_accuracy_compatible_kernel=true时，走的是和torch对齐的版本（lr和beta_pow都用double计算）。

### CODE REVIEW

File: python/paddle/optimizer/fusion_utils.py [hunk 1]
```diff
@@ -116,7 +116,8 @@ def build_buffer(self):
             if k not in self.accumulators_meta:
                 self.accumulators_meta[k] = {}
             for para_name, var_tmp in v.items():
-                assert var_tmp.dtype == self.dtype
+                if var_tmp.dtype != self.dtype:
+                    continue
                 src_len = var_tmp._numel() + get_align(var_tmp)
```
Comment (xxyux) [inline python/paddle/optimizer/fusion_utils.py:120]: 我们可以以fp32的格式存储beta1和beta2。
删除119-120行，并替换121行：
`src_len = (var_tmp._numel() + get_align(var_tmp)) * align[var_tmp.dtype] // align[self.dtype]`

File: python/paddle/optimizer/fusion_utils.py
Comment (xxyux) [inline python/paddle/optimizer/fusion_utils.py:184]: 在mapping的时候用src的dtype还原回去
`self.buffer._slice(start, end).view(src.dtype)`

Comment (xxyux) [inline python/paddle/optimizer/fusion_utils.py:188]: self.buffer._slice(start, end).view(src.dtype)._share_buffer_to(src)

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Review by yuanlehome — APPROVED** [review]

**Review by lugimzzz — APPROVED** [review]
lgtm

**Review by zrr1999 — APPROVED** [review]

**Review by zhwesky2010 — APPROVED** [review]
lgtm

**Review by From00 — APPROVED** [review]
